lib/elrequest.js

```diff
@@ -0,0 +1,187 @@
+      var connectTimer = setTimeout(function onConnectTimeout () {
+        // log.trace({url: url}, 'connectTimeout')
+        req.emit('connectTimeout')
+      }, connectTimeout)
+
+      socket.on('connect', function () {
+        // log.trace({url: url}, 'socket connected, clear connectTimeout')
+        clearTimeout(connectTimer)
+        connectTimer = null
+      })
+      socket.on('close', function () {
+        if (connectTimer) {
+          // log.trace({url: url}, 'socket close with active connectTimer, clear connectTimeout')
+          clearTimeout(connectTimer)
+        }
+      })
+    })
+  }
+
+  return req
+}
+
+module.exports = {
+  elRequest: elRequest
+}
+
+// ---- mainline
+
+// This main is only intended to demonstrate usage of this lib; not to be a
+// useful tool.
+//
+// Example:
+//      node elrequest.js https://www.elastic.co 100 1000 > elastic.html
+//
+// Example: download a ~1GB file, requiring a 30ms connect time, 2s idle timeout
+//      node elrequest.js http://mirror.uoregon.edu/ubuntu-releases/20.04/ubuntu-20.04.1-live-server-amd64.iso 30 2000 > ubuntu.iso
+//
+// Example: connect timeout because google port 81 drops TCP SYN packets:
+//      node elrequest.js http://www.google.com:81/foo 1000 10000
+//
+// Example: use `NODE_DEBUG=*` to see internal node debugging details
+//      NODE_DEBUG=* node elrequest.js $url 30 1000
+//
+// Example: quick connection, slow response
+//      % cat server.js
+//      var http = require('http');
+//      http.createServer(function (req, res) {
+//          console.log('SERVER: got request')
+//          setTimeout((function() {
+//              res.writeHead(200, {'Content-Type': 'text/plain'});
+//              res.write('line one\n')
+//              setTimeout((function() {
+//                  res.write('line two\n')
+//                  setTimeout((function() {
+//                      res.write('line three\n')
+//                      res.end()
+//                      console.log('SERVER: responded')
+//                  }), 500);
+//              }), 500);
+//          }), 500);
+//      }).listen(8080);
+//      % node server.js &
+//      % node elrequest.js http://127.0.0.1:8080/ 10 1000
+function main (argv) {
+  if (argv.length !== 5) {
+    process.stderr.write('elrequest: error: incorrect number of args\n')
+    process.stderr.write('usage: elrequest $url $connectTimeoutMs $timeoutMs\n')
+    process.exitCode = 1
+    return
+  }
+  const url = argv[2]
+  const connectTimeout = Number(argv[3])
+  const timeout = Number(argv[4])
+
+  var req = elRequest(url, {
+    timeout: timeout,
+    connectTimeout: connectTimeout
+    // TODO: log support
+  }, function onRes (res) {
+    res.pipe(process.stdout)
+  })
+
+  req.on('timeout', function () {
+    console.warn(`elrequest: response timeout (${timeout}ms): destroying request`)
+    req.destroy()
+    process.exitCode = 28 // using cURL's errno for a timeout
+  })
+
+  req.on('connectTimeout', function () {
```

Comment:
Dovetailing with the previous question about `connectTimeout` -- I presume that listening for this event _is_ required in order to end the full request due to our "socket ping" failing?

Reply:
Yes, if the user passes in a `opts.connectTimeout`, then they'll have to respond to that event to cancel/abort/destroy the request. (This roughly mirrors the behaviour of the "timeout" option and event.)